internal/pkg/scaffold/metrics_test.go

```diff
@@ -0,0 +1,97 @@
+// Copyright 2019 The Operator-SDK Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package scaffold
+
+import (
+	"testing"
+
+	"github.com/operator-framework/operator-sdk/internal/util/diffutil"
+)
+
+func TestMetrics(t *testing.T) {
+	r, err := NewResource(appApiVersion, appKind)
+	if err != nil {
+		t.Fatal(err)
+	}
+	s, buf := setupScaffoldAndWriter()
+	err = s.Execute(appConfig, &Metrics{Resource: r})
+	if err != nil {
+		t.Fatalf("Failed to execute the scaffold: (%v)", err)
+	}
+
+	if metricsExp != buf.String() {
+		diffs := diffutil.Diff(metricsExp, buf.String())
+		t.Fatalf("Expected vs actual differs.\n%v", diffs)
+	}
+}
+
+const metricsExp = `package metrics
+
+import (
+	"github.com/operator-framework/operator-sdk/pkg/k8sutil"
+	kubemetrics "github.com/operator-framework/operator-sdk/pkg/kube-metrics"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/client-go/rest"
+	ksmetric "k8s.io/kube-state-metrics/pkg/metric"
+	logf "sigs.k8s.io/controller-runtime/pkg/runtime/log"
+)
+
+var (
+	log        = logf.Log.WithName("metrics")
+	resource   = "app.example.com/v1alpha1"
+	kind       = "AppService"
+	metricName = "appservice_info"
+
+	MetricFamilies = []ksmetric.FamilyGenerator{
+		ksmetric.FamilyGenerator{
+			Name: metricName,
+			Type: ksmetric.Gauge,
+			Help: "Information about the AppService operator replica.",
+			GenerateFunc: func(obj interface{}) *ksmetric.Family {
+				crd := obj.(*unstructured.Unstructured)
+
+				return &ksmetric.Family{
+					Metrics: []*ksmetric.Metric{
+						{
+							Value:       1,
+							LabelKeys:   []string{"namespace", "appservice"},
+							LabelValues: []string{crd.GetNamespace(), crd.GetName()},
+						},
+					},
+				}
+			},
+		},
+	}
+)
+
+func ServeOperatorSpecificMetrics(cfg *rest.Config, host string, port int32) error {
```

Comment:
The `ServeOperatorSpecificMetrics()` function seems like it could be provided by in the SDK, which would make it easier for us to change its internals and patch bugs. See my main review comment for more of my thought process.

Reply:
Would we then still want to generate this file? The only thing that would be left is the above variables, would that make `main.go` file too cluttered if we decide to move it there?

My reasoning behind the metrics package was also that we only have the information about the api and kind name when the user adds the `api` and `controller`, so if we decide to move the leftover to main.go file that would mean updating the file after it would be generated. Let me know your thoughts.

Reply:
Yeah, I think we'd still generate this file, but we'd repurpose it to just be the abstraction between `main.go` and all of the API metrics. Similar to `pkg/apis/apis.go`, we could create this during `operator-sdk new` to setup a global variable and/or function that can be used to collect all of the metrics registrations for all of the APIs that get created by `operator-sdk add api`. This would serve basically the exact purpose that `pkg/apis/apis.go` serves, but for metrics instead of APIs.

Then the leftover variables at the top would go somewhere like `pkg/metrics/group/version/kind.go`, since they're specfiic to a GVK.

Thoughts?

Reply:
I see what you meant, so this would also cover a case where user adds multiple apis.

My only issue is generating so many files it might confuse the user and be a lot to mantain when user wants to update the name of the api, or just will skip updating, but I can see it being a cleaner solution. 

But having one global variable of all the collectors in a similiar way that is done in kube-state-metrics sgtm:
https://github.com/kubernetes/kube-state-metrics/blob/f6d391316ccd8291d8111fc2a7b2481b0eb376b0/internal/collector/builder.go#L93-L139